numpy/core/src/umath/ufunc_object.c

```diff
@@ -991,33 +992,45 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
  */
 static int
 check_for_trivial_loop(PyUFuncObject *ufunc,
-                        PyArrayObject **op,
-                        PyArray_Descr **dtype,
-                        npy_intp buffersize)
+        PyArrayObject **op, PyArray_Descr **dtypes,
+        npy_intp buffersize)
 {
-    npy_intp i, nin = ufunc->nin, nop = nin + ufunc->nout;
+    int i, nin = ufunc->nin, nop = nin + ufunc->nout;
 
     for (i = 0; i < nop; ++i) {
         /*
          * If the dtype doesn't match, or the array isn't aligned,
          * indicate that the trivial loop can't be done.
          */
-        if (op[i] != NULL &&
-                (!PyArray_ISALIGNED(op[i]) ||
-                !PyArray_EquivTypes(dtype[i], PyArray_DESCR(op[i]))
-                                        )) {
+        if (op[i] == NULL) {
+            continue;
+        }
+        int must_copy = !PyArray_ISALIGNED(op[i]);
+
+        if (dtypes[i] != PyArray_DESCR(op[i])) {
+            NPY_CASTING safety = PyArray_GetCastSafety(
+                    PyArray_DESCR(op[i]), dtypes[i], NULL);
+            if (safety < 0) {
+                /* A proper error during a cast check should be rare */
+                return -1;
+            }
+            if (!(safety & _NPY_CAST_IS_VIEW)) {
+                must_copy = 1;
+            }
```

Comment:
We don't actually have to check casting here right now.  The type resolver is currently in charge of checking cast safety.

Reply:
I'm confused by the explanation of what's not needed. 

But just looking at the code: I find the `must_copy` if not aligned or if a casting view is not possible a much clearer indication of what's happening than the `EquivTypes` before

Reply:
Yeah, nevermind.  My comment was that you may expect checks for calls like `np.add(a, b, out=a, casting="safe")` where `a` might not be a valid output based on the casting level provided.

In the later changes I actually move that check to here (I think).  Right now the is indeed `Equiv` just a less clear spelling of "view works".